libraries/standard/mqtt/src/iot_mqtt_operation.c

```diff
@@ -301,6 +332,158 @@ static bool _scheduleNextRetry( _mqttOperation_t * pOperation )
 
 /*-----------------------------------------------------------*/
 
+static bool _sendKeepAlivePingReq( _mqttConnection_t * pMqttConnection )
+{
+    bool status = true;
+    uint32_t swapStatus = 0;
+    size_t bytesSent = 0;
+    _mqttOperation_t * pPingreqOperation = &( pMqttConnection->pingreq );
+
+    /* Swap status is not checked when asserts are disabled. */
+    ( void ) swapStatus;
+
+    /* Because PINGREQ may be used to keep the MQTT connection alive, it is
+     * more important than other operations. Bypass the queue of jobs for
+     * operations by directly sending the PINGREQ in this job. */
+    bytesSent = pMqttConnection->pNetworkInterface->send( pMqttConnection->pNetworkConnection,
+                                                          pPingreqOperation->u.operation.pMqttPacket,
+                                                          pPingreqOperation->u.operation.packetSize );
+
+    if( bytesSent != pPingreqOperation->u.operation.packetSize )
+    {
+        IotLogError( "(MQTT connection %p) Failed to send PINGREQ.", pMqttConnection );
+        status = false;
+    }
+    else
+    {
+        /* Assume the keep-alive will fail. The network receive callback will
+         * clear the failure flag upon receiving a PINGRESP. */
+        swapStatus = Atomic_CompareAndSwap_u32( &( pPingreqOperation->u.operation.periodic.ping.failure ),
+                                                1,
+                                                0 );
+        IotMqtt_Assert( swapStatus == 1 );
+
+        /* Schedule a check for PINGRESP. */
+        pPingreqOperation->u.operation.periodic.ping.nextPeriodMs = IOT_MQTT_RESPONSE_WAIT_MS;
+
+        IotLogDebug( "(MQTT connection %p) PINGREQ sent. Scheduling check for PINGRESP in %d ms.",
+                     pMqttConnection,
+                     IOT_MQTT_RESPONSE_WAIT_MS );
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static IotMqttError_t _scheduleCallback( _mqttOperation_t * pOperation )
+{
+    IotMqttError_t status = IOT_MQTT_SCHEDULING_ERROR;
+    _mqttConnection_t * pMqttConnection = pOperation->pMqttConnection;
+
+    /* Non-waitable operation should have job reference of 1. */
+    IotMqtt_Assert( pOperation->u.operation.jobReference == 1 );
+
+    /* Schedule an invocation of the callback. */
+    if( pOperation->u.operation.notify.callback.function != NULL )
```

Comment:
Is it allowed for the app to specify a NULL pointer here? That seems like it should be a hard error.

Reply:
Yes, the user-provided callback function isn't required, and the callback struct is set to 0 in the [initializer](https://docs.aws.amazon.com/freertos/latest/lib-ref/c-sdk/mqtt/iot__mqtt__types_8h.html#a9776f60b74d9f1a80dd4036085ab1569)